### ISSUE TITLE
`kl_div` will crash in the backward pass on cuda

### ISSUE THREAD

**Issue by upupyy** [pytorch/pytorch#78867]
### 🐛 Describe the bug

`kl_div` will crash in the backward pass on cuda

```python
import torch

input = torch.rand([5, 5, 5], dtype=torch.float64, device='cuda', requires_grad=True)
target = torch.rand([5], dtype=torch.float64, device='cuda', requires_grad=True)

res = torch.nn.functional.kl_div(input, target)
print('forward')

res.sum().backward()
# crash
print('backward')
```

### Versions

pytorch: 1.11.0
cuda: 11.3

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @ngimel

**Comment by jbschlosser** [comment]
This is related to the differing shapes between `input` and `target`. Interestingly, an error message *is* thrown if `input` has 3 dims and `target` has 2:

```python
input = torch.rand([5, 5, 5], dtype=torch.float64, device='cuda', requires_grad=True)
target = torch.rand([5, 5], dtype=torch.float64, device='cuda', requires_grad=True)
torch.nn.functional.kl_div(input, target).sum().backward()
```

```
RuntimeError: output with shape [5, 5] doesn't match the broadcast shape [5, 5, 5]
```

**Comment by nikitaved** [comment]
These shapes are supposed to be supported, right? If so, then not only is it a shape issue, it is also an autograd issue. When the grads prop back to target, we need to analyze the shape of `input`  to figure out which dimensions to reduce over to make a grad of shape `input` to match the grad of shape `target`.

**Comment by nikitaved** [comment]
Note, however, `binary_cross_entropy` does not broadcast `target` over `input`:
```python
In [1]: import torch                                                                                                                                                         
                                                                                                                                                                             
In [2]:                                                                                                                                                                      
   ...: input = torch.rand([2, 2, 2], dtype=torch.float64, device='cpu', requires_grad=True)                                                                                 
   ...: target = torch.rand([2], dtype=torch.float64, device='cpu', requires_grad=True)                              
                                                                                                                                                                             
In [3]: torch.autograd.gradcheck(lambda i, t: torch.nn.functional.binary_cross_entropy_with_logits(i, t), [input, target])                                                   
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)                                                                                                  
Input In [3], in <cell line: 1>()                                                                                                                                            
----> 1 torch.autograd.gradcheck(lambda i, t: torch.nn.functional.binary_cross_entropy_with_logits(i, t), [input, target])

File ~/git/Quansight/pytorch/torch/autograd/gradcheck.py:1418, in gradcheck(func, inputs, eps, atol, rtol, raise_exception, check_sparse_nnz, nondet_tol, check_undefined_grad, check_grad_dtypes, check_batched_gr
ad, check_batched_forward_grad, check_forward_ad, check_backward_ad, fast_mode)                          
   1416         return False                        
   1417 else:                                       
-> 1418     return _gradcheck_helper(**args)                                                             

File ~/git/Quansight/pytorch/torch/autograd/gradcheck.py:1427, in _gradcheck_helper(func, inputs, eps, atol, rtol, check_sparse_nnz, nondet_tol, check_undefined_grad, check_grad_dtypes, check_batched_grad, check
_batched_forward_grad, check_forward_ad, check_backward_ad, fast_mode)                                   
   1424 tupled_inputs = _as_tuple(inputs)           
   1425 _check_inputs(tupled_inputs, check_sparse_nnz)                                                   
-> 1427 func_out = func(*tupled_inputs)             
   1428 outputs = _differentiable_outputs(func_out)                                                      
   1429 _check_outputs(outputs)                     

Input In [3], in <lambda>(i, t)                     
----> 1 torch.autograd.gradcheck(lambda i, t: torch.nn.functional.binary_cross_entropy_with_logits(i, t), [input, target])                                                                                         

File ~/git/Quansight/pytorch/torch/nn/functional.py:3104, in binary_cross_entropy_with_logits(input, target, weight, size_average, reduce, reduction, pos_weight)                                                  
   3101     reduction_enum = _Reduction.get_enum(reduction)                                              
   3103 if not (target.size() == input.size()):                                                          
-> 3104     raise ValueError("Target size ({}) must be the same as input size ({})".format(target.size(), input.size()))                                                                                           
   3106 return torch.binary_cross_entropy_with_logits(input, target, weight, pos_weight, reduction_enum)                                                                                                           

ValueError: Target size (torch.Size([2])) must be the same as input size (torch.Size([2, 2, 2]))                           
```

**Comment by Lezcano** [comment]
IMO losses should not broadcast but very often we do not impose this constraint explicitly and we may end up with some parts of the implementation allowing broadcasting, and somes not, which is what's happening here cc @jbschlosser

**Comment by nikitaved** [comment]
Why shouldn't they? The normalization is then has to change to the `numel` of a broadcasted shape...

**Comment by jbschlosser** [comment]
Agree that losses shouldn't broadcast in general - broadcasting is unlikely to match what the user wants to compute conceptually (e.g. broadcasting the same target across an entire batch is probably not correct and similarly vice-versa). Disallowing broadcasting for losses helps ensure the correct thing is being computed conceptually.

For this case, I wonder if it's possible to tighten up the restriction that `input.shape == target.shape` without breaking anything useful in practice.

**Comment by VitalyFedyunin** [comment]
@suo Can you please take a look

**Comment by jbschlosser** [comment]
@Lezcano / @nikitaved are looking into this, should be solved in #69212
